index.c: skip some strcmp() in case entry length is short enough.

mykaul opened the pull request:
In most cases we'll calculate - strlen() the length of the entry. Use that length - only if it's longer than 42, which is 'dirty-<uuid in ascii...>', then there's a point in checking if it matches 'dirty-...' or 'xattrop-...' string.

(Note: unsure, but I suspect we could have and should have skipped . and .. in as well)

Updates: #1000
Signed-off-by: Yaniv Kaul <ykaul@redhat.com>

Commit:
index.c: skip some strcmp() in case entry length is shortv enough.

In most cases we'll calculate - strlen() the length of the entry.
Use that length - only if it's longer than 42, which is 'dirty-<uuid in ascii...>',
then there's a point in checking if it matches 'dirty-...' or 'xattrop-...' string.

(Note: unsure, but I suspect we could have and should have skipped . and .. in as well)

Updates: #1000
Signed-off-by: Yaniv Kaul <ykaul@redhat.com>

mykaul commented:
/run regression

xhernandez commented on xlators/features/index/src/index.c:
```diff
@@ -523,17 +523,22 @@ index_fill_readdir(fd_t *fd, index_fd_ctx_t *fctx, DIR *dir, off_t off,
-                            strlen(DIRTY_SUBDIR "-"))) {
-            check_delete_stale_index_file(this, entry->d_name, DIRTY_SUBDIR);
-            continue;
+        entry_dname_len = strlen(entry->d_name);
+        if (entry_dname_len >= 42) {
+            /* 42 = strlen('dirty-') + UUID_CANONICAL_FORM_LEN (36) */
+            if (!strncmp(entry->d_name, XATTROP_SUBDIR "-",
+                         strlen(XATTROP_SUBDIR "-"))) {
+                check_delete_stale_index_file(this, entry->d_name,
+                                              XATTROP_SUBDIR);
+                continue;
+            } else if (!strncmp(entry->d_name, DIRTY_SUBDIR "-",
+                                strlen(DIRTY_SUBDIR "-"))) {
+                check_delete_stale_index_file(this, entry->d_name,
+                                              DIRTY_SUBDIR);
+                continue;
+            }
```
Actually the format of the name is very strict, it's just a fixed prefix, a `-`, and an uuid, so the length *must* be equal to 37 + length of prefix. If it's shorter or longer, it's not the desired entry.

We can define this function:
```c
static bool
index_check_index_file(const char *name, uint32_t nlen, const char *prefix)
{
    uint32_t plen;

    plen = strlen(prefix);
    if (nlen != plen + UUID_CANONICAL_FORM_LEN + 1) {
        return false;
    }

    if ((name[plen] != '-') || (memcmp(name, prefix, plen) != 0)) {
        return false;
    }

    check_delete_stale_index_file(this, name, prefix);

    return true;
}
```

And then rewrite this as:
```suggestion
        entry_dname_len = strlen(entry->d_name);
        if (index_check_index_file(entry->d_name, entry_dname_len,
                                   XATTROP_SUBDIR) ||
            index_check_index_file(entry->f_name, entry_dname_len,
                                   DIRTY_SUBDIR)) {
            continue;
        }
```

mykaul replied:
I've done it slightly different, which I think is OK too - and cheap enough for the effort - comparison with either lengths and that's it.
(I do wonder if we should skip '.' and '..' as well, but that's in a differnet patch)

xhernandez reviewed (commented):
There are other potential improvements in how entry names are manipulated, but they can be done in another patch if necessary.

Commit:
Fix review comments

mykaul commented:
/run regression

gluster-ant commented:
1 test(s) failed 
./tests/bugs/distribute/bug-1286171.t

0 test(s) generated core 


5 test(s) needed retry 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t
./tests/basic/afr/sparse-file-self-heal.t
./tests/basic/distribute/file-create.t
./tests/basic/shd-mux-ec.t
./tests/bugs/distribute/bug-1286171.t
https://build.gluster.org/job/gh_centos7-regression/2953/

Commit:
bit-rot-stub-helpers.c: skip checking for stale entries based on file name length.

skip some strcmp() in case entry length is short enough.

In most cases we'll calculate - strlen() the length of the entry.
Use that length - only if it's not 41, which is 'stub-<uuid in ascii...>',
then there's a point in checking if it matches 'stub-...'  string.

Updates: #1000
Signed-off-by: Yaniv Kaul <ykaul@redhat.com>

mykaul commented:
@xhernandez  - added similar code to bitrot xlator, please re-review.

mykaul commented:
/run regression

xhernandez commented on xlators/features/index/src/index.c:
```diff
@@ -523,17 +523,30 @@ index_fill_readdir(fd_t *fd, index_fd_ctx_t *fctx, DIR *dir, off_t off,
+         * or
+         * SLEN('xattrop') + '-' + UUID_CANONICAL_FORM_LEN (36)
+         */
+        if (entry_dname_len ==
+                (SLEN(XATTROP_SUBDIR) + 1 + UUID_CANONICAL_FORM_LEN) ||
+            entry_dname_len ==
+                (SLEN(DIRTY_SUBDIR) + 1 + UUID_CANONICAL_FORM_LEN)) {
+            if (!strncmp(entry->d_name, XATTROP_SUBDIR "-",
+                         strlen(XATTROP_SUBDIR "-"))) {
+                check_delete_stale_index_file(this, entry->d_name,
+                                              XATTROP_SUBDIR);
+                continue;
+            } else if (!strncmp(entry->d_name, DIRTY_SUBDIR "-",
+                                strlen(DIRTY_SUBDIR "-"))) {
+                check_delete_stale_index_file(this, entry->d_name,
+                                              DIRTY_SUBDIR);
+                continue;
+            }
```
I think this makes it harder to maintain if we need to add or remove indexes sometime, it's a bit less clear because of the multiple line splits, and does unnecessary comparisons for "dirty" entries, but it's fine enough.

mykaul replied:
Agree - we can certanily at some point improve readability. Small array with length and string to compare against, etc...
